### PR TITLE
Add `Convert match to let-else` assist

### PR DESCRIPTION
**Pull request by unexge**
Closes https://github.com/rust-lang/rust-analyzer/issues/13254

### CODE REVIEW

File: crates/hir-ty/src/infer.rs [hunk 1]
```diff
@@ -333,6 +333,8 @@ pub struct InferenceResult {
     assoc_resolutions: FxHashMap<ExprOrPatId, AssocItemId>,
     pub diagnostics: Vec<InferenceDiagnostic>,
     pub type_of_expr: ArenaMap<ExprId, Ty>,
+    /// For each match expr, record diverging arm's expr.
+    pub diverging_arms: FxHashMap<ExprId, Vec<ExprId>>,
```
Comment (unexge) [inline crates/hir-ty/src/infer.rs:337]: I've tried to get diverging arms by looking types of arm expressions via `hir::semantics::Semantics::type_of_expr` but we coerce all match arm expressions and the whole match expression to the `Never` type when we encounter any diverging arm, so I couldn't find a way to distinguish diverging and non-diverging arms without recording this information.

Comment (jonas-schievink) [reply crates/hir-ty/src/infer.rs:337]: Does it help if you look at the unadjusted type? In an expression like

```rust
    match Some(0) {
        Some(_) => {}
        None => return
    }
```

`return` has type `!` but is adjusted to `()`, so the unadjusted type should provide the information you need

Comment (jonas-schievink) [reply crates/hir-ty/src/infer.rs:337]: Ah, I found the issue – your tests use `Option` but don't declare it. I'll push a fix.

Comment (unexge) [reply crates/hir-ty/src/infer.rs]: ohh 😅 great! Thanks for the fix.

**Comment by fmease** [comment]
Does this also fix #11908?

**Comment by liamhession** [comment]
@unexge - I was hoping you could explain to me what you mean by the terms "diverging arm" and "extracting arm" in match statements? I'm not seeing that term used elsewhere so i'm not sure what exactly they're supposed to describe. The reason i'm interested is that i'm trying to expand the functionality of this convert-match-to-let-else assist that you created!

My current guess is that a diverging arm is one which simply `return`s? Extracting arm is everything else?

Would really appreciate any more context you may be able to give about how you approached creating this assist, thanks!

**Comment by Veykril** [comment]
diverging arm = a match arm that has type `!` (never type), so an arm that never yields a value (due to it `return`ing for example).

extracting I imagine is the arm that "extracts" the value of interest from the value being match on

**Comment by unexge** [comment]
@liamhession I believe @Veykril answered your question but to give a bit more context: 
There is a common theme that you are only interested some subpattern from a pattern, like you have `Option<T>` but you are only interested with `Some(T)` or you have [`ast::Expr`](https://github.com/rust-lang/rust-analyzer/blob/2656303c83c6949f4c0b8cb8058095dd48eee5da/crates/syntax/src/ast/generated/nodes.rs#L1513-L1546) but you are only interested with `ast::AwaitExpr` and the usual way to encode this pattern is you `match` on the given value (and now you can also use `let-else` statement :)) and you usually have 2 match arms:
one that diverges, meaning stops the execution of the function by `return`, `continue`, `panic!` or any other never `!` type 
one that extracts the value you are interested.

You can see that pattern a lot in Rust Analyzer especially in [`ide-assists`](https://github.com/rust-lang/rust-analyzer/tree/master/crates/ide-assists/src/handlers) for example:
```rust
// This only applies for tuple patterns, types, and initializers.
let tuple_pat = match pat { // <- extracts `ast::Pat::TuplePat` from `ast::Pat`
    ast::Pat::TuplePat(pat) => pat, // <- extracting arm
    _ => return None, // <- diverging arm
};
// ...
let tuple_init = match init { // <- extracts `ast::Expr::TupleExpr` from `ast::Expr`
    ast::Expr::TupleExpr(expr) => expr, // <- extracting arm
    _ => return None, // <- diverging arm
};
```
from https://github.com/rust-lang/rust-analyzer/blob/26562973b3482a635416b2b663a13016d4d90e20/crates/ide-assists/src/handlers/unwrap_tuple.rs#L33-L45

Happy to help if you have any questions and I can also try to give you some pointers if you have some specific improvements in your mind!

**Comment by liamhession** [comment]
Very nice context to have, thanks @unexge and @Veykril 

My confusion came about in part because i couldn't find really any other use of those terms "extracting arm" and "diverging arm" through Googling. But their names make sense in the context you laid out.

The improvements i'm aiming for are to expand the variety of "scrutinee expressions" (just learned that vocab from https://doc.rust-lang.org/reference/expressions/match-expr.html) this assist can be used for. The relevant issue within rust-analyzer is https://github.com/rust-lang/rust-analyzer/issues/13540

I've first been working on tweaking it so that it will recognize a statement with tuple after the `let` keyword, as outlined in the issue. But i hope to leave it even more general-purpose than just something that additionally recognizes that case.

I think first i'll need to update the `find_binding` function to add arms to the match statement it uses, and then update `rename_variable` to be able to use alternate sub-types of `ast::Pat`. Do you think that sounds about right @unexge ?

**Comment by unexge** [comment]
@liamhession yes I believe, you need to update following functions to be more generic:
`find_binding`: in addition to `ast::IdentPat` it should also find `ast::TuplePat` and other patterns that makes sense
`find_extracted_variable`: should probably return `Vec<ast::Name>` for every binding? instead of a single name
`rename_variable`: should be updated according to return types of previous functions and should handle new cases

**Pull request merged**